### DISCUSSION TITLE
ترجمة مصطلحي render و render function

### DISCUSSION THREAD

**Discussion by Abdelaziz18003** [vuejs-translations/docs-ar#133] [مصطلحات] [answered]
الترجمة المبدئية التي اعتمدناها هي "تصيير"، ومثلها فعل فريق React. المشكلة في هذه الترجمة هي أنها تحمل معنى غير مكتمل لغويا في هذا السياق، التصيير هو مصدر من الفعل "صيّر" وهو فعل متعد لمفعولين لايتم معناه إلا بهما: كأن نقول مثلا: صيّر الخباز الدقيقَ خبزًا، فإذا حولنا الفعل إلى مصدر قلنا "تصيير الدقيق خبزًا"، ولا يتم المعنى بخذف المفعول الثاني ونسكت كأن نقول "تصيير الخبزِ"، تصييره ماذا؟
كلمة render تحتمل هذا المعنى الذي هو معنى "جَعَلَ" أو "صَيَّرَ" كأن نقول "The new evidence rendered the old theory obsolete" والترجمة العربية الحرفية هنا يمكن أن تكون "الدليل الجديد جعل/صير النظريةَ القديمةَ باطلةََ"، لكنها في المقابل تحمل معان أخرى على حسب السياق الذي ترد فيه،
في سياق Vue.js، المقصود بال rendering هو إنشاء عناصر HTML قابلة للعرض على المتصفح إنطلاقا من مكونات vue
في سياق المتصفح أيضا ال Rendering Engine يعني البرنامج الذي يرسم عناصر ال HTML على الشاشة
عند النظر إلى هذه السياقات أرى أن كلمة "عرض" أصح وأنسب للسياق من كلمة "تصيير"
بهذا:
- كلمة Rendering تترجم إلى "عرض"
- كلمة Render Function تصبح "دالة العرض"
- كلمة Conditional Rendering تصبح "العرض الشرطي"
 
![image](https://github.com/vuejs-translations/docs-ar/assets/11301627/96a98518-8282-4d1f-8cfa-984ce7657f85)
![image](https://github.com/vuejs-translations/docs-ar/assets/11301627/0cd5dea6-c63f-4555-b4af-e7e7fef2e750)

**Comment by boussadjra** [comment]
شكرا على فتح المناقشة أخي عبد العزيز، هذا المصطلح يتكرر كثيرا في التوثيق، وقد فكرت مليا في ترجمته بشكل صحيح ولم أفلح في إيجاد ترجمة دقيقة  ولكن اعتمدت كلمة تصيير لأنه سبق استعمالها رغم أنها ليست مقنعة بشكل كافي. في المقابل كلمة عرض تتجاوز المعنى وهي تدل في هذه الحالة أن المحتوى صار مرئيا (shown/displayed) أو بشكل أدق مرسوما على المخرج النهائي (الشاشة مثلا)، فعلى سبيل إذا تحدثنا عن الـSSR فنقول العرض من جانب الخادوم وهذا يوهمنا بأن هناك نتيجة مرئية بينما في الحقيقة الأمر يتعلق بتصريف شيفرة التطبيق وانتاج محتوى يرسل في حال طُلب من جانب العميل حيث يُعرض في النهاية.
فحسب وجهة نظري فإن كلمة تصيير لا توفي بالمعنى حتى في الاستعمال الانجليزي وكلمة عرض تتجاوزه، ولكن إذا دققنا جيدا فإن كلمة render يمكن ترجمتها بشكل مختلف حسب السياق فمثلا: 
- نقول العرض الشرطي إذا تحدثنا عن conditional rendering أو عرض القوائم بالنسبة لـv-for
- دالة تصنيع العقد الافتراضية بالنسبة لـrender function 
<img width="688" alt="render-pipeline" src="https://github.com/vuejs-translations/docs-ar/assets/11801238/ff333ed7-5b40-4db2-bcf5-5bb580e714f3">
 فكما ترى في مسار هذا المخطط، دور هذه الدالة بعيد عن أي عملية عرض.

**Comment by Abdelaziz18003** [comment]
أولا أرى أننا لقد فعلنا الشيء الصحيح حين بدأنا الترجمة دون تدقيق شديد في المصطلحات وإلا لما تقدمنا بهذا الشكل الكبير لو بقينا ندقق مع كل مصطلح، المرحلة القادمة في نظري هي مراجعة المصطلحات وهذا ما بدأت فيه بالفعل، إذ بدأت القراءة من البداية وكلما يصادفني مصطلح لا أقتنع بترجمته سأفتح نقاشا حوله، بعدما ندقق المصطلحات ستأتي مرحلة التدقيق اللغوي بحيث نعيد صياغة العبارات تحريا للفصيح من اللغة بدلا من الترجمة الحرفية. أرى أننا نسير في الطريق الصحيح.
بالنسبة لكلمة render فأنا أتفق معك في أنها لا تفيد العرض الفعلي على الشاشة في سياق render function خلافا ل conditional rendering التي نحن متفقان عليها، أرى أن دالة تصنيع العقد الإفتراضية ترجمة جيدة وتعبر عن المعنى، قلقي الوحيد حولها هو أنها سوف تختصر فيما بعد إلى "دالة التصنيع" وقد يظنها السامع أو القارئ ترجمة ل Factory Function، أرى لو نستعين ببعض الآراء الأخرى من مجتمع المطورين المهتمين بالترجمة ونجمع قائمة من المقترحات نتخير منها ما نراه الأنسب.

**المقترحات الحالية:**

- **دالة العرض**: ترجمة مختصرة لكنها لا تعبر عن المعنى الدقيق في هذا السياق
- **دالة تصنيع العقد الإفتراضية**: دقيقة لكنها طويلة نوعا ما ويمكن أن تفهم كترجمة ل Factory function اذا ما اختصرت فيما بعد إلى "دالة التصنيع"
- **دالة تخريج العقد الإفتراضية**: أضيفها كمقترح ثالث ويمكن اختصارها إلى دالة التخريج دون أن تختلط بمعنى آخر فيما أعلم.
- **دالة التصريف**: (صرَّف الأشياءَ :نقلها، بدَّلها، وجَّهها، وصرَّف العُملةَ: حوَّلها وبدَّلها بمثلها) فأرى أن التصريف يمكن أن يصدق على المعنى المراد هنا لأنه تحويل لعناصر vue إلى عناصر HTML إفتراضية.

**Reply by itsmolaboune** [reply]
**دالة تخريج العقد الافتراضية": أنسب  ترجمة في سياق Virtual DOM rendering وهذا ماتقوم به Vue.

**Comment by boussadjra** [comment]
أرى أن كلمة تخريج أفضل من تصريف التي تستعمل غالبا تستعمل لترجمة compilation.
بالنسبة للتدقيق والتصحيح سأطور مكون عبارة عن رابط يوجه القارئ إلى القضية الأصلية في  المستودع حيث يمكنه التعليق واعطاء ملاحظاته.

**Reply by ousszizou** [reply]
كلمة **تخريج**  أقرب و أنسب لما يحدث فعليا في هاته المرحلة.

**Comment by WatheqAlshowaiter** [comment]
شكرا على هذا النقاش.. 

أرى "دالة العرض" جيدة

**Reply by WatheqAlshowaiter** [reply]
في هذه التغريدة اقتراحات كثيرة للترجمة 
https://twitter.com/watheq_show/status/1646213584106422297

**Reply by boussadjra** [reply]
في سياق Vue ترجمتها إلى عبارة "دالة العرض" لا يعكس دورها الحقيقي والذي هو انتاج أو تصنيع العقد الافتراضية الخاص بالـDOM الافتراضي، ومنه نستنتج أن لا علاقة لها بالعرض، فعندما نقول عرض فإننا نقصد بها بشكل بديهي شيء قابل للرؤية.
وأرى أن عبارات دالة التخريج، الترجمة أو الإخراج هي أقرب للمعنى. ويمكن اسقاط نفس المعنى على ترجمة SSR، بينما يختلف الأمر عندما نتحدث عن مصطلح renderless والتي يمكن ترجمتها بمصطلح "عديم العرض"

**Reply by bRIMOs** [reply]
ماذا عن  دالة التجسيد في هذه الحالة

**Reply by Abdelaziz18003** [reply]
@bRIMOs هل يمكنك التفصيل أكثر عن الرابط الذي يجعل كلمة "تجسيد" مناسبة للمعنى المقصود هنا؟

**Comment by boussadjra** [comment]
كخلاصة ماذا تقترحون لترجمة هاته العبارات:
- Rendering (عرض) 
- Render function (دالة تخريج)
- Conditional rendering (العرض الشرطي)
- Server side rendering (التخريج من جانب الخادم)
- Renderless components (دالة عديمة المخرج)
- Renderer (مُصيِّر، عارض)
 
العبارات التي بين قوسين بعضها من اقتراحي والبعض من اقتراح الإخوة.

**Reply by Abdelaziz18003** [reply]
حسنا، هذه العبارات التي لدي تعقيب عليها، ما لم أذكره هنا فأنا أتفق معك فيه:
 
- كلمة Rendering : أرى أن لها ترجمتان، عرض وتخريج، تستعملان حسب السياق
- عبارة Server side rendering : أرى اختصارها قليلا إلى "التخريج على الخادم" أو ما شابهها
- عبارة Renderless components: الحديث هنا عن الكونات التي لها منطق بينما لا تعرض شيئا، أقترح "مكون عديم الواجهة"
- كلمة Renderer: أرى أن نحافظ على الإتساق هنا، ونستعمل اسم الفاعل لكلا الكلمتين السابقتين حسب السياق، (عارض، مُخرِّج)

**Comment by muhammetsait** [comment]
أود اقتراح بعض الترجمات الأخرى:

كلمة render الإنكليزية معناها الأصلي (وهو إلقاء عمل موسيقي أو مسرحي على الجمهور) يمكن ترجمته إلى (تمثيل) أو (تجسيد)

المعنى المعتمد في سياق الحوسبة عامة هو (إنتاج الصورة النهائية لمشهد ثلاثي الأبعاد) أو (تحويل الكود إلى صورة وأشكال) فربما صلح استخدام كلمة (تصوير) أو (تشكيل) للتعبير عنه

Render function (دالة تصوير)
Conditional rendering (التصوير الشرطي)
Server side rendering (التصوير على الخادم)
Renderless components (مكوّن بلا صورة)
Renderer (دالة/جهة التصوير)

Render function (دالة تشكيل)
Conditional rendering (التشكيل الشرطي)
Server side rendering (التشكيل على -جهة- الخادم)
Renderless components (مكوّن دون تشكيل)
Renderer (دالة/جهة التشكيل)

**Reply by Abdelaziz18003** [reply]
شكرا @muhammetsait على هذا الاقتراح، صراحة أعجبتني كلمة "تشكيل"، أرى أنها بسيطة وتعبر عن المعنى كما أنها غير مستعملة في التعبير عن مفاهيم أخرى في البرمجة حسب علمي، فلا يلتبس معناها بمعان أخرى، ما رأيك يا @boussadjra

**Reply by boussadjra** [reply]
نعم كلمة تشكيل جميلة وتفي بالمعنى لترجمة بعض المصطلحات وليس كلها، فمثلا في الـconditional rendering يقصد بها العرض على الشاشة وبالتالي كلمة العرض جيدة في هذا السياق، أما مصطلح render function  فيصلح ترجمته بدالة التشكيل وقس على ذلك في باقي المصطلحات

**Reply by Abdelaziz18003** [reply, marked as answer]
أتفق معك تماما، سنستبدل "تصيير" ب "تشكيل"، أما render التي تعني معنى الإظهار (الذي يقابله الإخفاء) فسنبقي على مصطلح العرض للتعبير عنها.
وبهذا يكون لدينا مصطلحان يقابلان كلمة "render":

1. "تشكيل" يستعمل في التعبير عن "render" التي تعني تحويل العنصر إلى عقد افتراضية
2. "عرض" يستعمل في التعبير عن السياقات التي تعني الإظهار على الشاشة

أرى أننا بهذا قد وفقنا في ترجمة هذه الكلمة "الصعبة"، سأعتبر هذه المناقشة منتهية، شكرا لكم جميعا على المساهمة في إثرائها.
